core/src/main/java/org/apache/iceberg/TableScanContext.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.util.Collection;
+import java.util.Map;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.expressions.Expressions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+
+/**
+ * Context object with optional arguments for a TableScan.
+ */
+final class TableScanContext {
+  private final Long snapshotId;
+  private final Expression rowFilter;
+  private final boolean ignoreResiduals;
+  private final boolean caseSensitive;
+  private final boolean colStats;
+  private final Collection<String> selectedColumns;
+  private final ImmutableMap<String, String> options;
+  private final Long fromSnapshotId;
+  private final Long toSnapshotId;
+
+  TableScanContext() {
+    this(null, Expressions.alwaysTrue(), false, true, false, null, ImmutableMap.of(), null, null);
+  }
+
+  private TableScanContext(Long snapshotId, Expression rowFilter, boolean ignoreResiduals,
+                           boolean caseSensitive, boolean colStats, Collection<String> selectedColumns,
+                           ImmutableMap<String, String> options, Long fromSnapshotId, Long toSnapshotId) {
+    this.snapshotId = snapshotId;
+    this.rowFilter = rowFilter;
+    this.ignoreResiduals = ignoreResiduals;
+    this.caseSensitive = caseSensitive;
+    this.colStats = colStats;
+    this.selectedColumns = selectedColumns;
+    this.options = options;
+    this.fromSnapshotId = fromSnapshotId;
+    this.toSnapshotId = toSnapshotId;
+  }
+
+  Long snapshotId() {
+    return snapshotId;
+  }
+
+  TableScanContext snapshotId(Long id) {
```

Comment:
How about `scanSnapshotId`?

Also, why is `id` a boxed `Long`? If this is called, it should be non-null, right?

Reply:
> If this is called, it should be non-null, right?

No, it is set to null in the constructor of `IncrementalDataTableScan`: https://github.com/apache/iceberg/blob/e0180f783f62f4cd0f30b13a62ef5fcd1cc6f3b9/core/src/main/java/org/apache/iceberg/IncrementalDataTableScan.java#L44